cpp/lib/ClientConfigParser.cpp

```diff
@@ -0,0 +1,315 @@
+  const std::string SF_CLIENT_CONFIG_ENV_NAME("SF_CLIENT_CONFIG_FILE");
+  std::set<std::string> KnownCommonEntries = {"log_level", "log_path"};
+}
+
+////////////////////////////////////////////////////////////////////////////////
+void load_client_config(
+  const char* in_configFilePath,
+  client_config* out_clientConfig)
+{
+  ClientConfigParser configParser;
+  ClientConfig clientConfig;
+  configParser.loadClientConfig(in_configFilePath, clientConfig);
+  if (!clientConfig.logLevel.empty())
+  {
+    out_clientConfig->logLevel = (char*)SF_CALLOC(1, sizeof(clientConfig.logLevel));
+    sf_memcpy(out_clientConfig->logLevel, sizeof(out_clientConfig->logLevel),
+      clientConfig.logLevel.data(), clientConfig.logLevel.size());
+  }
+  if (!clientConfig.logPath.empty())
+  {
+    out_clientConfig->logPath = (char*)SF_CALLOC(1, sizeof(clientConfig.logPath));
+    sf_memcpy(out_clientConfig->logPath, MAX_PATH,
+      clientConfig.logPath.data(), clientConfig.logPath.size());
+  }
+}
+
+// Public ======================================================================
+////////////////////////////////////////////////////////////////////////////////
+ClientConfigParser::ClientConfigParser()
+{
+  ; // Do nothing.
+}
+
+////////////////////////////////////////////////////////////////////////////////
+ClientConfigParser::~ClientConfigParser()
+{
+  ; // Do nothing.
+}
+
+////////////////////////////////////////////////////////////////////////////////
+void ClientConfigParser::loadClientConfig(
+  const std::string& in_configFilePath,
+  ClientConfig& out_clientConfig)
+{
+  char envbuf[MAX_PATH + 1];
+  std::string derivedConfigPath = "";
+  if (!in_configFilePath.empty())
+  {
+    // 1. Try config file if it was passed in
+    derivedConfigPath = in_configFilePath;
+    CXX_LOG_INFO("sf", "ClientConfigParser", "loadClientConfig",
+      "Using client configuration path from a connection string: %s", in_configFilePath.c_str());
+  }
+  else if (const char* clientConfigEnv = sf_getenv_s(SF_CLIENT_CONFIG_ENV_NAME.c_str(), envbuf, sizeof(envbuf)))
+  {
+    // 2. Try environment variable SF_CLIENT_CONFIG_ENV_NAME
+    derivedConfigPath = clientConfigEnv;
+    CXX_LOG_INFO("sf", "ClientConfigParser", "loadClientConfig",
+      "Using client configuration path from an environment variable: %s", clientConfigEnv);
+  }
+  else
+  {
+    // 3. Try DLL binary dir
+    std::string binaryDir = getBinaryPath();
+    std::string binaryDirFilePath = binaryDir + SF_CLIENT_CONFIG_FILE_NAME;
+    if (checkFileExists(binaryDirFilePath))
+    {
+      derivedConfigPath = binaryDirFilePath;
+      CXX_LOG_INFO("sf", "ClientConfigParser", "loadClientConfig",
+        "Using client configuration path from binary directory: %s", binaryDirFilePath.c_str());
+    }
+    else
+    {
+#if defined(WIN32) || defined(_WIN64)
+      // 4. Try user home dir
+      std::string homeDir = sf_getenv_s("USERPROFILE", envbuf, sizeof(envbuf));
```

Comment:
Since the whole function is just to resolve the client config, I think it might make more sense to just keep it all there (I'll move all the resolving code to a separate function from loadClientConfig instead)